Bottom: 16c4290b99405851896a4e579e320b83d432d4fb
Top:    283449168669ab42abc2b19afd0fc06e5cf2ab74
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-11-24 19:54:37 +0000

Refresh of try-yet-another-way

---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index 4c0469d..936b8bf 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -874,95 +874,96 @@ hidgl_fill_pcb_polygon (PolygonType *poly, const BoxType *clip_box)
       return;
     }
 
-  if (poly->Clipped->contour_tree->size > 1) {
-
-    /* Polygon has holes */
-    debug_holey_polygon_count++;
-
-    /* XXX: Need to assert() that there is room to increment the stencil counter value */
-
-    /* Flush out any existing geoemtry to be rendered */
-    hidgl_flush_triangles (&buffer);
-
-    /* Save the stencil setup */
-    glPushAttrib (GL_STENCIL_BUFFER_BIT);
-
-    /*   ___________________________________________________
-     *  |   __________                                      |
-     *  |  |          |                                1x0  |
-     *  |  | 0x0      |                                     |
-     *  |  |      ....|....                                 |
-     *  |  |      :   |   :                                 |
-     *  |  |______:___|   :                                 |
-     *  |         :.......: <-Hole we're about to mask      |
-     *  |___________________________________________________|
-     *
-     * NEXT: Mask out the holes:
-     *
-     *    Set the object masking bit for regions we don't want to draw.
-     *    We MUST reset this bit afterwards when drawing the outline.
-     *    IE. outlines must always enclose all holes.
-     */
-
-    glPushAttrib (GL_COLOR_BUFFER_BIT);
-    glColorMask (0, 0, 0, 0);
+  if (poly->Clipped->contour_tree->size <= 1) {
+    /* Polygon does not have holes */
+    fill_contour (poly->Clipped->contours);
+    return;
+  }
+
+  debug_holey_polygon_count++;
+
+  /* XXX: Need to assert() that there is room to increment the stencil counter value */
+
+  /* Flush out any existing geoemtry to be rendered */
+  hidgl_flush_triangles (&buffer);
+
+  /* Save the stencil setup */
+  glPushAttrib (GL_STENCIL_BUFFER_BIT);
+
+  /*   ___________________________________________________
+   *  |   __________                                      |
+   *  |  |          |                                1x0  |
+   *  |  | 0x0      |                                     |
+   *  |  |      ....|....                                 |
+   *  |  |      :   |   :                                 |
+   *  |  |______:___|   :                                 |
+   *  |         :.......: <-Hole we're about to mask      |
+   *  |___________________________________________________|
+   *
+   * NEXT: Mask out the holes:
+   *
+   *    Set the object masking bit for regions we don't want to draw.
+   *    We MUST reset this bit afterwards when drawing the outline.
+   *    IE. outlines must always enclose all holes.
+   */
+
+  glPushAttrib (GL_COLOR_BUFFER_BIT);
+  glColorMask (0, 0, 0, 0);
 
 //    glStencilMask (layer_stencil_bit | 1);
 //    glStencilFunc (GL_EQUAL, layer_stencil_bit, layer_stencil_bit);
-    glStencilOp (GL_KEEP, GL_KEEP, GL_INCR);
-    r_search (poly->Clipped->contour_tree, clip_box, NULL, do_hole, NULL);
-    hidgl_flush_triangles (&buffer);
-
-    glPopAttrib ();
-
-    /*   ___________________________________________________
-     *  |   __________                                      |
-     *  |  |    ......|...........                     1x0  |
-     *  |  | 0x0: 0x0 |          :                          |
-     *  |  |    :  ...|...       :                          |
-     *  |  |    : :0x0|   :      :                          |
-     *  |  |____:_:___|1x1:      :                          |
-     *  |       : :.......:      :                          |
-     *  |       :            1x0 :                          |
-     *  |       :................: <-Polygon outer          |
-     *  |___________________________________________________|
-     *
-     * NEXT: Draw the polygon:
-     *
-     *    We draw our polygon where the stencil test passes
-     *    Where we draw we GL_REPLACE the stencil buffer with 1x0
-     */
+  glStencilOp (GL_KEEP, GL_KEEP, GL_INCR);
+  r_search (poly->Clipped->contour_tree, clip_box, NULL, do_hole, NULL);
+  hidgl_flush_triangles (&buffer);
+
+  glPopAttrib ();
+
+  /*   ___________________________________________________
+   *  |   __________                                      |
+   *  |  |    ......|...........                     1x0  |
+   *  |  | 0x0: 0x0 |          :                          |
+   *  |  |    :  ...|...       :                          |
+   *  |  |    : :0x0|   :      :                          |
+   *  |  |____:_:___|1x1:      :                          |
+   *  |       : :.......:      :                          |
+   *  |       :            1x0 :                          |
+   *  |       :................: <-Polygon outer          |
+   *  |___________________________________________________|
+   *
+   * NEXT: Draw the polygon:
+   *
+   *    We draw our polygon where the stencil test passes
+   *    Where we draw we GL_REPLACE the stencil buffer with 1x0
+   */
 
 //    glStencilFunc (GL_EQUAL, layer_stencil_bit, layer_stencil_bit | 1);
 
 
-    /* XXX: DAMN, The stencil op operates on the whole word, thus causing breakage */
-    glStencilOp (GL_DECR, GL_KEEP, GL_ZERO);
-    fill_contour (poly->Clipped->contours);
-    hidgl_flush_triangles (&buffer);
-
-    /*   ___________________________________________________
-     *  |   __________                                      |
-     *  |  |     .....|__________                      1x0  |
-     *  |  | 0x0: 0x0 |          |                          |
-     *  |  |    :  ...|___       |                          |
-     *  |  |    : :0x0|1x0|      |                          |
-     *  |  |____:_:___|   |      |                          |
-     *  |       | |_______|      |                          |
-     *  |       |           0x0  |                          |
-     *  |       |________________|                          |
-     *  |___________________________________________________|
-     *
-     * DONE (just cleanup left)
-     */
-
-    /* Restore the stencil buffer setup */
-    glPopAttrib ();
+  /* XXX: DAMN, The stencil op operates on the whole word, thus causing breakage */
+  glStencilOp (GL_DECR, GL_KEEP, GL_ZERO);
+  fill_contour (poly->Clipped->contours);
+  hidgl_flush_triangles (&buffer);
+
+// 11111xxxxx[1|0]
+// 
+
+  /*   ___________________________________________________
+   *  |   __________                                      |
+   *  |  |     .....|__________                      1x0  |
+   *  |  | 0x0: 0x0 |          |                          |
+   *  |  |    :  ...|___       |                          |
+   *  |  |    : :0x0|1x0|      |                          |
+   *  |  |____:_:___|   |      |                          |
+   *  |       | |_______|      |                          |
+   *  |       |           0x0  |                          |
+   *  |       |________________|                          |
+   *  |___________________________________________________|
+   *
+   * DONE (just cleanup left)
+   */
 
-  } else {
-    /* Polygon does not have holes */
-    fill_contour (poly->Clipped->contours);
-  }
+  /* Restore the stencil buffer setup */
+  glPopAttrib ();
 }
 
 void
